### ISSUE TITLE
APIs for MPI-like functionality?

### ISSUE THREAD

**Issue by jiamings** [ray-project/ray#842]
I wonder if it there are efficient APIs (or implementations) for MPI-like functionality? For example, perform reduce sum very efficiently on some list of objects, which points to numpy arrays.

While this can be achieved with the current API (e.g. in the tutorial), but maybe having additional APIs can prevent reinventing wheels. 

Or maybe just doing np.sum() in a single thread is good enough?

**Comment by pcmoritz** [comment]
Hey Tony,

Which APIs are you especially interested in, besides the reduce or all reduce and what is the setting you are interested in (number of nodes, number of objects, data sizes)?

We are trying to keep the API minimal and as you point out this functionality can be implemented with the current API (either ray.get the objects on a single node and do np.sum or implement a tree reduce with the API). However, we have been thinking about adding an all-reduce like primitive for a use case we are interested in, which is averaging neural network weights for training/SGD (if that can lead to nontrivial performance improvements).

If you end up trying out to do the operations you are interested in with the current API and run into bottlenecks, let us know! That information will be very useful when we implement a solution.

-- Philipp.

**Comment by programmerlwj** [comment]
Hey,

I'm using the mpi4py API like AllReduce, Bcast etc,  and I think these are really easy to use.  For example, I can use the gather and reduce within the actor (actually is the actor with rank 0). 

However, it seems that gather and reduce APIs should be outside the actors if I use ray. So I have to define another class except the actor's class to do gather or broadcast,  is that right?

Thanks.

--Wanjun

**Comment by jiamings** [comment]
I am also considering the use case of averaging neural network weights or gradients for updates; not entirely sure whether it would lead to performance improvements over the current API; but if it does it could be really helpful. 

Looking forward to a solution!

**Comment by robertnishihara** [comment]
Supporting MPI primitives is largely an API question. One possible API that could allow people to port MPI code pretty easily would be something like

```python
# This corresponds to a usual MPI worker.
def mpi_worker_code():
    # do some stuff
    ray.mpi.allreduce(...)
    # do some stuff
    if ray.mpi.rank() == 0:
        # do some stuff
    ...

# Start a bunch of "Ray MPI" workers.
ray.mpi_run(mpi_worker_code, num_mpi_workers=10)
```

Under the hood, `ray.mpi_run` would start a number of actors that all have handles to one another and can send and receive. There are some subtleties with the implementation however.

cc @elibol
